cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -117,6 +120,130 @@ inline std::shared_ptr<std::unordered_map<std::pair<StreamId, VersionId>, AtomKe
+    }
+
+    void do_react(const pipelines::SnapshotVersionQuery&) {
+        util::raise_rte("Snapshot not currently supported in generic batch read version");
+    }
+};
+
+inline std::optional<AtomKey> get_key_for_version_query(const std::shared_ptr<VersionMapEntry>& version_map_entry, const pipelines::VersionQuery& version_query) {
+    return util::variant_match(version_query.content_,
+        [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) {
+        return find_index_key_for_version_id(specific_version.version_id_, version_map_entry);
+    },
+    [&version_map_entry] (const pipelines::TimestampVersionQuery& timestamp_version) {
+       return find_index_key_for_version_timestamp(timestamp_version.timestamp_, version_map_entry);
+    },
+    [&version_map_entry] (const std::monostate&) {
+       return version_map_entry->get_first_index(false);
+       },
+    [] (const auto&) -> std::optional<AtomKey> {
+        util::raise_rte("Unsupported version query type");
+    });
+}
+
+inline std::vector<folly::Future<std::optional<AtomKey>>> batch_get_versions(
+    const std::shared_ptr<Store>& store,
+    const std::shared_ptr<VersionMap>& version_map,
+    const std::vector<StreamId>& symbols,
+    const std::vector<pipelines::VersionQuery>& version_queries) {
+    ARCTICDB_SAMPLE(BatchGetVersion, 0)
+    util::check(symbols.size() == version_queries.size(), "Symbol and version query list mismatch: {} != {}", symbols.size(), version_queries.size());
+
+    robin_hood::unordered_flat_map<StreamId, StreamVersionData> version_data;
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        auto it = version_data.find(*symbol);
+        if(it == version_data.end())
+            version_data.insert(robin_hood::pair<StreamId, StreamVersionData>(*symbol, StreamVersionData{version_queries[symbol.index]}));
+        else
+            it->second.react(version_queries[symbol.index]);
+    }
+
+    using SplitterType = folly::FutureSplitter<std::shared_ptr<VersionMapEntry>>;
+    robin_hood::unordered_flat_map<StreamId, SplitterType> shared_futures;
+    std::vector<folly::Future<std::optional<AtomKey>>> output;
+    output.reserve(symbols.size());
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        const auto it = version_data.find(*symbol);
+        util::check(it != version_data.end(), "Missing version data for symbol {}", *symbol);
+        auto version_entry_fut = folly::Future<std::shared_ptr<VersionMapEntry>>::makeEmpty();
+        if(it->second.count_ == 1) {
+            version_entry_fut = async::submit_io_task(CheckReloadTask{store, version_map, *symbol, it->second.load_param_});
+        } else {
+            auto fut = shared_futures.find(*symbol);
+            if(fut == shared_futures.end()) {
+                auto [splitter, inserted] = shared_futures.emplace(*symbol, folly::FutureSplitter(async::submit_io_task(CheckReloadTask{store, version_map, *symbol, it->second.load_param_})));
```

Comment:
@vasil-pashov moving your comment here https://github.com/man-group/ArcticDB/pull/423/files#r1205063759

Reply:
From the time you commented it, I refactored all this code and I think now this function looks much clearer. Regarding this line in particular. As for your comment, I also splatted the line in two, to improve the clarity.

Reply:
This line looks very similar to before and is still too long!

I urge you to extract the `async::submit_io_task(...)` to a temporary variable and `std::move` its result here.

Reply:
sorry @qc00, it is exactly what I already did. But my push command went wrong and I didn't realise of that